### PR TITLE
cream-bugfix: CA-209507 and CA-209508

### PR DESCRIPTION
**Pull request by johnelse**

### CODE REVIEW

**Comment by simonjbeaumont** [comment]
I'm on the hook for reviewing this apparently.

Note to self: yhis is quite a large back backport of two issues from trunk. There are two additional commits:
- 6d21e45: this allowed for a much cleaner backport
- 844463b: this is because the code is different in cream from trunk

Apart from these there should be a 1:1 mapping between commits here and commits in the original pull-requests[1]. I'm planning to review only the ones that conflicted during backport and the additional commits mentioned above.

[1]: https://github.com/xapi-project/xen-api/pull/2650 and https://github.com/xapi-project/xen-api/pull/2657

File: ocaml/xapi/message_forwarding.ml [hunk 1]
```diff
@@ -922,8 +922,12 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
 						(fun () ->
 							finally_clear_host_operation ~__context ~host:suitable_host ?host_op ();
 							(* In certain cases, VM might have been destroyed as a consequence of operation *)
-							if Db.is_valid_ref __context vm then
-								Db.VM.set_scheduled_to_be_resident_on ~__context ~self:vm ~value:Ref.null))
+							if Db.is_valid_ref __context vm then begin
+								Db.VM.set_resident_on ~__context ~self:vm ~value:Ref.null;
+								List.iter
+									(fun vgpu -> Db.VGPU.set_scheduled_to_be_resident_on ~__context ~self:vgpu ~value:Ref.null)
```
Comment (simonjbeaumont) [inline ocaml/xapi/message_forwarding.ml]: This isn't relevant in cream and makes some of the future backported patches misleading.

**Comment by simonjbeaumont** [comment]
Just a change to be made to the backport https://github.com/xenserver/xen-api/pull/111#discussion-diff-68229775 to remove the field that doesn't exist in Cream. I realise when you've done this all the commits will be rebased but I'm happy apart from that one change.

**Comment by johnelse** [comment]
Thanks @simonjbeaumont

**Pull request merged**